src/protocol/sort/shuffle.rs

```diff
@@ -103,70 +119,121 @@ impl<'a, F: Field, S: Step + SpaceIndex> Shuffle<'a, F, S> {
         try_join_all(reshares).await
     }
 
-    /// `single_shuffle` is called for the helpers
-    /// i)   2 helpers generate random sequence using shared prss random number
+    /// `single` is called for the helpers
+    /// i)   2 helpers receive permutation pair and choose the permutation to be applied
     /// ii)  2 helpers apply the permutation to their shares
     /// iii) reshare to `to_helper`
-    #[allow(clippy::cast_possible_truncation)]
-    async fn single_shuffle<M: Mesh, G: Gateway<M, S>>(
+    async fn single<M: Mesh, G: Gateway<M, S>>(
         &mut self,
+        shuffle_or_unshuffle: ShuffleOrUnshuffle,
         ctx: &ProtocolContext<'_, G, S>,
         which_step: ShuffleStep,
+        permutations: &(Permutation, Permutation),
     ) -> Result<Vec<Replicated<F>>, BoxError> {
         let to_helper = Self::shuffle_for_helper(which_step);
         let step = (self.step_fn)(which_step);
-        let prss = &ctx.participant[step];
         let channel = ctx.gateway.get_channel(step);
 
         if to_helper != channel.identity() {
-            let direction = if to_helper.peer(Direction::Left) == channel.identity() {
-                Direction::Left
+            let permute = if to_helper.peer(Direction::Left) == channel.identity() {
+                &permutations.0
             } else {
-                Direction::Right
+                &permutations.1
             };
-            let mut permute = Self::generate_random_permutation(self.input.len(), direction, prss);
-            apply_inv(&mut permute, &mut self.input);
+
+            if matches!(shuffle_or_unshuffle, ShuffleOrUnshuffle::Shuffle) {
```

Comment:
Is this the canonical way to do this in Rust? Is it not to do a two line match expression with => operators?